### PR TITLE
Avoid returning default from MethodSymbol.Parameters

### PR DESCRIPTION
**Pull request by sharwell**
Review commit-by-commit recommended.

This change ensures that all implementations of `MethodSymbol.Parameters` return a non-default array, excluding cases where `LazyMethodChecks` fails to initialize parameters. It also cleans up some initialization sites to use current practices.

### CODE REVIEW

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 4)

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For the second review.

File: src/Compilers/CSharp/Portable/Lowering/SynthesizedMethodBaseSymbol.cs [hunk 1]
```diff
@@ -106,7 +106,7 @@ public sealed override ImmutableArray<ParameterSymbol> Parameters
             {
                 if (_parameters.IsDefault)
                 {
-                    ImmutableInterlocked.InterlockedCompareExchange(ref _parameters, MakeParameters(), default(ImmutableArray<ParameterSymbol>));
+                    ImmutableInterlocked.InterlockedInitialize(ref _parameters, MakeParameters());
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Lowering/SynthesizedMethodBaseSymbol.cs:109]: Can you help me understand how these two lines are different? I'm missing the distinction here.

Comment (sharwell) [reply src/Compilers/CSharp/Portable/Lowering/SynthesizedMethodBaseSymbol.cs:109]: ➡️ They are semantically identical. The simpler form allows a reader to understand the behavior without the mental burden of relating the third parameter of `InterlockedCompareExchange` to the default state for the field.

**Review by jaredpar — APPROVED** [review]

**Comment by jcouv** [comment]
Merged/squashed. Will take another look at https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1341051 with this new baseline. Thanks

**Comment by sharwell** [comment]
@jcouv Please do not _squash_ merge pull requests submitted by me. I meticulously separated this pull request into four distinct changes with the intent they would be preserved as such. Thanks!

**Comment by jcouv** [comment]
@sharwell Sorry about that.

**Comment by sharwell** [comment]
@jcouv no problem glad it's merged !

**Pull request merged**